server/example/workflow-example/src/main/java/com/eclipsesource/glsp/example/workflow/utils/WorkflowBuilder.java

```diff
@@ -15,245 +15,181 @@
-				GModelUtil.generateId(node, genId, modelState);
-			} else {
-				node.setId(id);
-			}
-			node.setType(type);
-			node.setPosition(position);
-			node.setSize(size);
-			return node;
-		}
-
-		abstract public T build();
 	}
 
-	public static class ActivityNodeBuilder extends GNodeBuilder<ActivityNode> {
+	public static class ActivityNodeBuilder extends AbstractGNodeBuilder<ActivityNode, ActivityNodeBuilder> {
 		protected String nodeType;
 
-		public ActivityNodeBuilder(GraphicalModelState modelState, String type, String nodeType) {
-			super(modelState, type);
+		public ActivityNodeBuilder(String type, String nodeType) {
+			super(type);
 			this.nodeType = nodeType;
 		}
 
-		public ActivityNode build() {
-			ActivityNode activityNode = WfgraphFactory.eINSTANCE.createActivityNode();
-			fillData(activityNode, "activityNode");
-			activityNode.setNodeType(nodeType);
-			return activityNode;
+		@Override
+		protected void setProperties(ActivityNode node) {
+			super.setProperties(node);
+			node.setNodeType(nodeType);
+		}
+
+		@Override
+		protected ActivityNode instantiate() {
+			return WfgraphFactory.eINSTANCE.createActivityNode();
+		}
+
+		@Override
+		protected ActivityNodeBuilder self() {
+			return this;
 		}
 	}
 
-	public static class TaskNodeBuilder extends GNodeBuilder<TaskNode> {
+	public static class TaskNodeBuilder extends AbstractGNodeBuilder<TaskNode, TaskNodeBuilder> {
 		private String name;
 		private String taskType;
 		private int duration;
 
-		public TaskNodeBuilder(GraphicalModelState modelState, String type, String name, String taskType,
+		public TaskNodeBuilder(String type, String name, String taskType,
 				int duration) {
-			super(modelState, type);
+			super(type);
 			this.name = name;
 			this.taskType = taskType;
 			this.duration = duration;
+
 		}
 
-		public TaskNode build() {
-			TaskNode taskNode = WfgraphFactory.eINSTANCE.createTaskNode();
-			fillData(taskNode, "task");
+		@Override
+		protected TaskNode instantiate() {
+			return WfgraphFactory.eINSTANCE.createTaskNode();
+		}
+
+		@Override
+		protected TaskNodeBuilder self() {
+			return this;
+		}
+
+		@Override
+		public void setProperties(TaskNode taskNode) {
+			super.setProperties(taskNode);
 			taskNode.setName(name);
 			taskNode.setTaskType(taskType);
 			taskNode.setDuration(duration);
 			taskNode.setLayout("vbox");
 			taskNode.getChildren().add(createCompartment(taskNode));
-			return taskNode;
 		}
 
 		private GCompartment createCompartment(TaskNode taskNode) {
-			GCompartment compHeader = GraphFactory.eINSTANCE.createGCompartment();
-			compHeader.setType(ModelTypes.COMP_HEADER);
-			compHeader.setId(taskNode.getId() + "_header");
-			compHeader.setLayout("hbox");
-
-			compHeader.getChildren().add(createCompartmentIcon(taskNode));
-			compHeader.getChildren().add(createCompartmentHeader(taskNode));
-			return compHeader;
+			return new GCompartmentBuilder().setType(taskType)
+					.setType(ModelTypes.COMP_HEADER)
```

Comment:
This looks wrong (two invocations of `setType`).

Reply:
Why don't we set the type via the builder's constructor? The abstract superclasses do that and we could overload the constructor. If no parameter value, the default type is used, otherwise we overwrite it with a specific type. This would spare an additional line. What do you think?

Reply:
Good point. I will add the constructor overloading for all leaf classes.